File: .tasks/2026-09-21-social-studio-handoff/checks/browser-check.mjs
import assert from 'node:assert/strict';
import fs from 'node:fs/promises';
import { existsSync, readdirSync } from 'node:fs';
import os from 'node:os';
import path from 'node:path';
import { fileURLToPath } from 'node:url';
import { chromium } from 'playwright-core';

// Runs against the offline reference, without product credentials or APIs.
const here = path.dirname(fileURLToPath(import.meta.url));
const reference = path.resolve(here, '../reference');
const screenshots = path.join(here, 'screenshots');
const cache = path.join(os.homedir(), '.cache/ms-playwright');
const cached = existsSync(cache)
  ? readdirSync(cache).filter(name => name.startsWith('chromium-')).sort()
      .flatMap(name => ['chrome-linux64/chrome', 'chrome-linux/chrome'].map(exe => path.join(cache, name, exe)))
  : [];
const executablePath = [process.env.BROWSER_PATH, '/usr/bin/chromium', '/usr/bin/chromium-browser', ...cached]
  .filter(Boolean).find(candidate => existsSync(candidate));
if (!executablePath) throw new Error('Set BROWSER_PATH to an installed Chromium executable.');
await fs.mkdir(screenshots, { recursive: true });
const html = await fs.readFile(path.join(reference, 'preview.html'), 'utf8');
const passed = [];
const widths = [];
const mark = name => { passed.push(name); console.log('PASS ' + name); };
const browser = await chromium.launch({ executablePath, headless: true, args: ['--no-sandbox'] });
try {
  const page = await browser.newPage({ viewport: { width: 1440, height: 1080 } });
  page.setDefaultTimeout(5000);
  const errors = [];
  let networkRequests = 0;
  page.on('pageerror', error => errors.push(error.message));
  page.on('request', () => { networkRequests += 1; });
  await page.route('**/*', route => route.abort());
  const settle = async () => {
    await page.evaluate(async () => {
      await document.fonts.ready;
      await Promise.all([...document.images].map(img => img.complete ? Promise.resolve() : new Promise(resolve => {
        img.addEventListener('load', resolve, { once: true });
        img.addEventListener('error', resolve, { once: true });
      })));
      await new Promise(resolve => requestAnimationFrame(() => requestAnimationFrame(resolve)));
    });
  };
  const fit = async view => {
    const measurement = await page.evaluate(() => ({ viewport: innerWidth, document: document.documentElement.scrollWidth }));
    assert.ok(measurement.document <= measurement.viewport, view + ': horizontal overflow');
    widths.push({ view, ...measurement });
  };
  const shot = async name => {
    await settle();
    await page.screenshot({ path: path.join(screenshots, name + '.png'), fullPage: false, animations: 'disabled' });
  };
  const count = async expected => page.waitForFunction(n => document.querySelectorAll('article').length === n, expected);
  const textIncludes = async (selector, text) => page.waitForFunction(
    ({ selector, text }) => document.querySelector(selector)?.textContent.includes(text), { selector, text });

  await page.setContent(html, { waitUntil: 'load' });
  await count(3);
  await fit('desktop-queue');
  assert.equal(await page.locator('.studio-shell').evaluate(el => getComputedStyle(el).backgroundColor), 'rgb(13, 13, 13)');
  assert.match(await page.locator('.btn-primary').first().evaluate(el => getComputedStyle(el).backgroundImage), /rgb\(197, 160, 89\)/);
  assert.equal(await page.locator('.btn-primary').first().evaluate(el => getComputedStyle(el).color), 'rgb(36, 26, 18)');
  assert.ok(await page.locator('aside .brand-logo').evaluate(img => img.complete && img.naturalWidth === 96));
  await shot('preview-desktop');
  mark('Desktop: three review cards, brand colors, dark text on gold and loaded community logo');

  for (const [tab, name] of [['autopilot', 'preview-autopilot'], ['templates', 'preview-templates'], ['accounts', 'preview-accounts']]) {
    await page.locator('#tab-' + tab).click();
    await page.evaluate(() => scrollTo(0, 0));
    await fit('desktop-' + tab);
    await shot(name);
    if (tab === 'autopilot') {
      await page.evaluate(() => scrollTo(0, document.documentElement.scrollHeight));
      await shot('preview-autopilot-bottom');
      await page.evaluate(() => scrollTo(0, 0));
    }
  }
  mark('Settled viewport screenshots for the four work areas');

  await page.locator('#tab-queue').click();
  await page.locator('[data-action="approve"][data-id="1"]').click();
  await count(2);
  await page.locator('[data-filter="scheduled"]').click();
  await count(3);
  mark('Approval moves one demo clip to the scheduled filter after successful save');

  await page.locator('[data-filter="all"]').click();
  await page.locator('#search').fill('Hook');
  await count(1);
  await page.locator('#search').fill('no such clip 92347');
  await page.getByText('Keine Clips gefunden', { exact: true }).waitFor({ state: 'visible' });
  await page.locator('#search').fill('');
  mark('Search and empty-state handling');
  await page.locator('[data-layout="grid"]').click();
  assert.equal(await page.locator('#queue-list').getAttribute('class'), 'grid-view');
  await page.locator('[data-layout="list"]').click();
  mark('Grid and list switch');

  await page.locator('[data-action="menu"][data-id="2"]').click();
  await page.getByRole('button', { name: 'Transkript bearbeiten', exact: true }).click();
  await page.locator('#dialog').waitFor({ state: 'visible' });
  for (let i = 0; i < 9; i += 1) {
    assert.ok(await page.evaluate(() => document.querySelector('#dialog').contains(document.activeElement)));
    await page.keyboard.press('Tab');
  }
  await page.keyboard.press('Escape');
  assert.equal(await page.locator('#dialog').isVisible(), false);
  assert.ok(await page.locator('#clip-more-2').evaluate(el => el === document.activeElement));
  mark('Dialog focus containment, Escape and return to trigger');

  await page.locator('#tab-accounts').click();
  await page.locator('[data-action="fail-next"]').click();
  await page.locator('#tab-queue').click();
  await page.locator('[data-filter="review"]').click();
  await page.locator('[data-action="approve"][data-id="2"]').click();
  await textIncludes('[data-clip="2"] [role="alert"]', 'nicht gespeichert');
  assert.equal(await page.locator('[data-clip="2"]').getAttribute('data-status'), 'review');
  mark('Failed approval retains the clip and shows a local error');

  await page.locator('#tab-autopilot').click();
  const weekly = page.locator('[name="youtube.posts_per_week"]');
  await weekly.fill('71');
  await page.locator('#schedule-form button[type="submit"]').click();
  assert.equal(await weekly.getAttribute('aria-invalid'), 'true');
  assert.ok(await weekly.evaluate(el => el === document.activeElement));
  await weekly.fill('5');
  await page.locator('#schedule-form button[type="submit"]').click();
  await textIncludes('#save-status', 'Gespeichert');
  mark('Invalid schedule values rejected, valid draft saved');

  await page.locator('#tab-accounts').click();
  await page.locator('[data-action="fail-next"]').click();
  await page.locator('#tab-autopilot').click();
  await page.locator('[name="youtube.posts_per_week"]').fill('6');
  await page.locator('#schedule-form button[type="submit"]').click();
  await textIncludes('#save-status', 'nicht gespeichert');
  assert.equal(await page.locator('[name="youtube.posts_per_week"]').inputValue(), '6');
  await page.locator('[data-action="reset-schedule"]').click();
  assert.equal(await page.locator('[name="youtube.posts_per_week"]').inputValue(), '5');
  mark('Failed save preserves draft; reset restores last successful demo state');

  await page.locator('#tab-queue').focus();
  for (const [key, id] of [['ArrowRight', 'tab-autopilot'], ['End', 'tab-accounts'], ['Home', 'tab-queue']]) {
    await page.keyboard.press(key);
    assert.ok(await page.locator('#' + id).evaluate(el => el === document.activeElement));
  }
  mark('Tabs support arrow keys, Home and End');

  await page.locator('#tab-templates').click();
  await page.locator('[data-action="template"]').first().click();
  await page.locator('#position-slider').fill('70');
  await page.locator('#position-slider').dispatchEvent('input');
  await shot('preview-editor');
  await page.locator('#layout-form button[type="submit"]').click();
  await page.locator('#dialog').waitFor({ state: 'hidden' });
  await page.locator('[data-action="template"]').first().click();
  assert.equal(await page.locator('#position-slider').inputValue(), '70');
  await page.keyboard.press('Escape');
  mark('Schematic layout demo saves and restores local settings');

  for (const width of [390, 320]) {
    await page.setViewportSize({ width, height: 844 });
    assert.ok(await page.locator('header .brand-logo').isVisible());
    for (const tab of ['queue', 'autopilot', 'templates', 'accounts']) {
      await page.locator('#tab-' + tab).click();
      await fit(width + '-' + tab);
    }
    await page.locator('#tab-queue').click();
    if (width === 390) {
      await page.evaluate(() => scrollTo(0, 0));
      await shot('preview-mobile');
      await page.locator('article').first().scrollIntoViewIfNeeded();
      await shot('preview-mobile-cards');
    }
  }
  mark('Four mobile work areas at 390px and 320px, loaded logo, no document overflow');
  assert.deepEqual(errors, []);
  assert.equal(networkRequests, 0);
  mark('No JavaScript errors or network requests; no product APIs used');

  await fs.writeFile(path.join(here, 'browser-results.json'), JSON.stringify({
    passed: passed.length, checks: passed, browser: browser.version(),
    target: '../reference/preview.html only', font_rendering: 'system fallback; not proof of product fonts',
    screenshots: 'new viewport captures on the project host, not original chat PNGs',
    js_errors: errors.length, network_requests: networkRequests, widths,
  }, null, 2) + '\n');
  console.log('Reference browser checks passed: ' + passed.length);
} finally {
  await browser.close();
}

File: .tasks/2026-09-21-social-studio-handoff/reference/tests/browser_check.py
"""Visual/reference-preview checks. Requires Python Playwright and Chromium.
This tests the offline HTML, not the separate React integration components.
"""
from pathlib import Path
from playwright.sync_api import sync_playwright, expect
import json

ROOT = Path(__file__).resolve().parents[1]
results = []

def passed(name):
    results.append(name)
    print('PASS', name)

with sync_playwright() as p:
    browser = p.chromium.launch(executable_path='/usr/bin/chromium', headless=True, args=['--no-sandbox'])
    page = browser.new_page(viewport={'width':1440, 'height':1080})
    page.set_default_timeout(4000)
    js_errors = []
    requests = []
    page.on('pageerror', lambda error: js_errors.append(str(error)))
    page.on('request', lambda request: requests.append(request.url))
    page.set_content((ROOT/'preview.html').read_text(), wait_until='load')
    expect(page.locator('article')).to_have_count(3)
    assert not page.evaluate('document.documentElement.scrollWidth > innerWidth')
    page.screenshot(path=str(ROOT/'preview-desktop.png'), full_page=True)
    passed('Desktop: 3 review cards, no horizontal overflow')
    assert page.locator('aside .brand-logo').evaluate('(img) => img.complete && img.naturalWidth === 96')
    assert page.locator('header .brand-logo').is_hidden()
    assert page.locator('.studio-shell').evaluate('(el) => getComputedStyle(el).backgroundColor') == 'rgb(13, 13, 13)'
    assert 'rgb(197, 160, 89)' in page.locator('.btn-primary').first.evaluate('(el) => getComputedStyle(el).backgroundImage')
    assert page.locator('.btn-primary').first.evaluate('(el) => getComputedStyle(el).color') == 'rgb(36, 26, 18)'
    passed('Compiled brand: neutral black, gold actions with dark text, authentic sidebar logo')

    page.locator('[data-action="approve"][data-id="1"]').click()
    expect(page.locator('article')).to_have_count(2)
    page.locator('[data-filter="scheduled"]').click()
    expect(page.locator('article')).to_have_count(3)
    passed('Approval moves one clip into scheduled; only after save')

    page.locator('[data-filter="all"]').click()
    page.locator('#search').fill('Hook')
    expect(page.locator('article')).to_have_count(1)
    page.locator('#search').fill('no such clip 92347')
    expect(page.get_by_text('Keine Clips gefunden', exact=True)).to_be_visible()
    passed('Search and empty-state handling')
    page.locator('#search').fill('')
    page.locator('[data-layout="grid"]').click()
    expect(page.locator('#queue-list')).to_have_class('grid-view')
    page.locator('[data-layout="list"]').click()
    passed('Card-grid and list views')

    page.locator('[data-action="menu"][data-id="2"]').click()
    expect(page.locator('#clip-menu')).to_be_visible()
    page.get_by_role('button', name='Transkript bearbeiten', exact=True).click()
    expect(page.locator('#dialog')).to_be_visible()
    assert page.evaluate('document.querySelector("#dialog").contains(document.activeElement)')
    for _ in range(9):
        page.keyboard.press('Tab')
        assert page.evaluate('document.querySelector("#dialog").contains(document.activeElement)')
    page.keyboard.press('Escape')
    expect(page.locator('#dialog')).not_to_be_visible()
    expect(page.locator('#clip-more-2')).to_be_focused()
    passed('Dialog contains focus; Escape returns focus to trigger')

    page.locator('#tab-accounts').click()
    page.locator('[data-action="fail-next"]').click()
    page.locator('#tab-queue').click()
    page.locator('[data-filter="review"]').click()
    page.locator('[data-action="approve"][data-id="2"]').click()
    expect(page.locator('[data-clip="2"] [role="alert"]')).to_contain_text('nicht gespeichert')
    expect(page.locator('[data-clip="2"]')).to_have_attribute('data-status','review')
    passed('Failed approval retains the clip and displays a local error')

    page.locator('#tab-autopilot').click()
    field = page.locator('[name="youtube.posts_per_week"]')
    field.fill('71')
    page.locator('#schedule-form button[type="submit"]').click()
    expect(field).to_have_attribute('aria-invalid','true')
    expect(field).to_be_focused()
    field.fill('5')
    page.locator('#schedule-form button[type="submit"]').click()
    expect(page.locator('#save-status')).to_contain_text('Gespeichert')
    passed('Autopilot: invalid numbers blocked, valid edits saved')

    page.locator('#tab-accounts').click()
    page.locator('[data-action="fail-next"]').click()
    page.locator('#tab-autopilot').click()
    page.locator('[name="youtube.posts_per_week"]').fill('6')
    page.locator('#schedule-form button[type="submit"]').click()
    expect(page.locator('#save-status')).to_contain_text('nicht gespeichert')
    expect(page.locator('[name="youtube.posts_per_week"]')).to_have_value('6')
    page.locator('[data-action="reset-schedule"]').click()
    expect(page.locator('[name="youtube.posts_per_week"]')).to_have_value('5')
    passed('Failed schedule save retains draft; reset restores last successful state')
    page.screenshot(path=str(ROOT/'preview-autopilot.png'), full_page=True)

    page.locator('#tab-queue').focus()
    page.keyboard.press('ArrowRight')
    expect(page.locator('#tab-autopilot')).to_be_focused()
    page.keyboard.press('End')
    expect(page.locator('#tab-accounts')).to_be_focused()
    page.keyboard.press('Home')
    expect(page.locator('#tab-queue')).to_be_focused()
    passed('Tabs: arrows, Home and End')

    page.locator('#tab-templates').click()
    page.locator('[data-action="template"]').first.click()
    page.locator('#position-slider').fill('70')
    page.locator('#position-slider').dispatch_event('input')
    page.screenshot(path=str(ROOT/'preview-editor.png'), full_page=True)
    page.locator('#layout-form button[type="submit"]').click()
    expect(page.locator('#dialog')).not_to_be_visible()
    page.locator('[data-action="template"]').first.click()
    expect(page.locator('#position-slider')).to_have_value('70')
    page.keyboard.press('Escape')
    passed('Focused layout demo opens, saves local settings and restores them')

    # Smaller layouts must scroll internally, not overflow the document.
    for width in [390, 320]:
        page.set_viewport_size({'width':width,'height':844})
        assert page.locator('header .brand-logo').is_visible()
        for tab in ['queue','autopilot','templates','accounts']:
            page.locator('#tab-'+tab).click()
            assert not page.evaluate('document.documentElement.scrollWidth > innerWidth'), (width,tab)
        page.locator('#tab-queue').click()
        if width == 390:
            page.screenshot(path=str(ROOT/'preview-mobile.png'), full_page=True)
    passed('All four areas fit 390px and 320px without document overflow')

    assert not js_errors, js_errors
    assert not requests, requests
    passed('No JavaScript errors, API calls or network requests in the standalone preview')
    browser.close()

(ROOT/'tests/browser-results.json').write_text(json.dumps({'passed':len(results),'checks':results,'target':'preview.html only', 'font_rendering':'standalone preview uses the system fallback; product stylesheet references existing self-hosted Manrope and Sora'},ensure_ascii=False,indent=2)+'\n')
